### PR TITLE
Split [Build.t] into [Build.t] and [Build.With_targets.t]

### PR DESCRIPTION
**Pull request by snowleopard**
This PR splits `Build.t` into `Build.t` and `Build.With_targets.t`, allowing us to distinguish build expressions that declare targets from those that do not at the type level. 

This touches quite a bit of code and in some situations introduces previously unnecessary conversions (from `Build.t` to  `Build.With_targets.t`), but I think the benefits (see below) outweigh the cost. 

Several invariants (and associated consistency checks) are removed:

* `Build.If_then_else` holds `Build.t` expressions that declare no targets.
* `Build.Lazy_no_target` holds `Build.t` expressions that declare no targets.
* `Build.Memo` holds `Build.t` expressions that declare no targets.
* `Commands.Args.Dyn` holds `Build.t` expressions that declare no targets.
* `Build_system.prefix_rules` previously assumed that the given `prefix` declares no targets.
* `Command.Expand` is used only with `Command.t` descriptions that declare no targets.

Because we no longer need to traverse `Build.t` expressions to compute targets, we save quite a lot of repeated work which was previously done in `Build.targets` which didn't implement any memoization of `Build.Memo` nodes. For example, when building Dune, we entered `Build.Memo` around 16000 times (in the call of `Build.targets`), whereas `Build.exec` that does proper memoization does this only ~1000 times.

Finally, in a couple of place we cached the `Build.targets` computation. This is no longer necessary.

Benchmarking on building Dune with Dune does not show any degradation in terms of performance or memory consumption.

### CODE REVIEW

**Comment by snowleopard** [comment]
Thanks for the review, @rgrinberg! I haven't actually tried the `Build.No_targets.t` variant. I'll check how well it works and if it's better than the current approach I'll switch to it.

File: src/dune/build.ml [hunk 1]
```diff
@@ -180,101 +180,124 @@ let source_tree ~dir =
...
+    add ~targets
+      ( match res with
+      | Ok x -> f x
+      | Error e ->
+        { build = Fail { fail = (fun () -> raise e) }
+        ; targets = Path.Build.Set.empty
+        } )
+end
+
+let add build ~targets : _ With_targets.t =
```
Comment (aalekseyev) [inline src/dune/build.ml]: Should this be called `Build.add_targets` or, perhaps, `Build.with_targets`? `add` makes me think of adding `Built.t` values together, or something like that.

Comment (snowleopard) [reply src/dune/build.ml]: Yep, it used to be `Build.with_targets ~targets` but it felt a bit repetitive. I agree that `add` is an ambiguous name, so will bring `with_targets` back.

File: src/dune/command.ml [hunk 2]
```diff
@@ -82,30 +72,43 @@ let expand_static_exn ~dir t =
...
+      Build.add ~targets:[ fn ]
+        (Build.return [ Path.reach (Path.build fn) ~from:dir ])
+    | Dyn dyn ->
+      Build.no_targets
+        (Build.dyn_deps (Build.map dyn ~f:(expand_static_exn ~dir)))
+    | Fail f -> Build.no_targets (Build.fail f)
+    | Hidden_deps deps ->
+      Build.no_targets (Build.map (Build.deps deps) ~f:(fun () -> []))
+    | Hidden_targets fns -> Build.add ~targets:fns (Build.return [])
+    | Expand f ->
```
Comment (aalekseyev) [inline src/dune/command.ml]: With the type of `expand` changing, the names `Expand` are a bit out of sync now. I don't have any good ideas though. `Expand_no_targets` could work, but we might as well keep `Expand` if you like that more.

Comment (snowleopard) [reply src/dune/command.ml]: I've updated the comment in `command.mli` to make it clearer that `expand` does not include target declarations.

Comment (aalekseyev) [reply src/dune/command.ml]: Looks good.

**Comment by snowleopard** [comment]
@aalekseyev Thanks! I've done some renaming:
* `Build.add` to `Build.with_targets`
* `Build.no_targets` to `Build.with_no_targets`

This seems like a clearer naming scheme.

@rgrinberg I've just checked: we use operators from `Build.O` about twice more often than operators from `Build.With_targets.O`, so I guess the current split is more lightweight compared to the alternative that you mentioned (`Build.t`/`Build.No_targets.t`). It seems best to stick to the current approach.

**Pull request merged**